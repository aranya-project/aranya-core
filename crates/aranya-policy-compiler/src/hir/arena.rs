use std::borrow::Cow;

use aranya_policy_ast::{self as ast, AstNode};
use serde::{Deserialize, Serialize};
use slotmap::SecondaryMap;

use crate::hir::hir::{
    ActionArgId, ActionId, BlockId, CmdFieldId, CmdId, EffectFieldId, EffectId, EnumId, ExprId,
    FactId, FactKeyId, FactValId, FinishFuncArgId, FinishFuncId, FuncArgId, FuncId, GlobalId,
    IdentId, StmtId, StructFieldId, StructId, VTypeId,
};

/// An arena of AST nodes.
#[derive(Clone, Default, Debug, Serialize, Deserialize)]
#[cfg_attr(test, derive(PartialEq))]
pub(crate) struct AstNodes<'ast> {
    pub actions: SecondaryMap<ActionId, Cow<'ast, AstNode<ast::ActionDefinition>>>,
    pub action_args: SecondaryMap<ActionArgId, Cow<'ast, ast::FieldDefinition>>,
    pub cmds: SecondaryMap<CmdId, Cow<'ast, AstNode<ast::CommandDefinition>>>,
    pub cmd_fields: SecondaryMap<CmdFieldId, Cow<'ast, ast::StructItem<ast::FieldDefinition>>>,
    pub effects: SecondaryMap<EffectId, Cow<'ast, AstNode<ast::EffectDefinition>>>,
    pub effect_fields:
        SecondaryMap<EffectFieldId, Cow<'ast, ast::StructItem<ast::EffectFieldDefinition>>>,
    pub enums: SecondaryMap<EnumId, Cow<'ast, AstNode<ast::EnumDefinition>>>,
    pub facts: SecondaryMap<FactId, Cow<'ast, AstNode<ast::FactDefinition>>>,
    pub fact_keys: SecondaryMap<FactKeyId, Cow<'ast, ast::FieldDefinition>>,
    pub fact_vals: SecondaryMap<FactValId, Cow<'ast, ast::FieldDefinition>>,
    pub finish_funcs: SecondaryMap<FinishFuncId, Cow<'ast, AstNode<ast::FinishFunctionDefinition>>>,
    pub finish_func_args: SecondaryMap<FinishFuncArgId, Cow<'ast, ast::FieldDefinition>>,
    pub funcs: SecondaryMap<FuncId, Cow<'ast, AstNode<ast::FunctionDefinition>>>,
    pub func_args: SecondaryMap<FuncArgId, Cow<'ast, ast::FieldDefinition>>,
    pub global_lets: SecondaryMap<GlobalId, Cow<'ast, AstNode<ast::GlobalLetStatement>>>,
    pub structs: SecondaryMap<StructId, Cow<'ast, AstNode<ast::StructDefinition>>>,
    pub struct_fields:
        SecondaryMap<StructFieldId, Cow<'ast, ast::StructItem<ast::FieldDefinition>>>,
    pub stmts: SecondaryMap<StmtId, Cow<'ast, AstNode<ast::Statement>>>,
    pub exprs: SecondaryMap<ExprId, Cow<'ast, ast::Expression>>,
    pub idents: SecondaryMap<IdentId, Cow<'ast, ast::Identifier>>,
    pub blocks: SecondaryMap<BlockId, Cow<'ast, Vec<AstNode<ast::Statement>>>>,
    pub types: SecondaryMap<VTypeId, Cow<'ast, ast::VType>>,
}
