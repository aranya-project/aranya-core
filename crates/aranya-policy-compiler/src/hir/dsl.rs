//! Provides a convenient macro for constructing [`Hir`] in
//! tests.

#![cfg(test)]

use slotmap::KeyData;

/// Creates an ID from an index and version using slotmap's KeyData.
///
/// # Panics
/// - If idx is 0
/// - If version is 0
pub(crate) fn make_id<T: From<KeyData>>(idx: u32, version: u32) -> T {
    assert!(idx > 0, "idx must be greater than 0");
    assert!(version > 0, "version must be greater than 0");
    let v = (idx as u64) | ((version as u64) << 32);
    KeyData::from_ffi(v).into()
}
