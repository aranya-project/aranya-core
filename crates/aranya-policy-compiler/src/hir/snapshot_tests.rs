#![cfg(test)]

use aranya_policy_ast::Version;
use aranya_policy_lang::lang::parse_policy_str;
use serde::Serialize;

use crate::hir::{self, arena::AstNodes, hir::Hir};

/// Wrapper struct for clean JSON snapshots
#[derive(Serialize)]
struct ParseSnapshot<'a> {
    hir: &'a Hir,
    ast_nodes: &'a AstNodes<'a>,
}

#[test]
fn test_parse_basic_policy() {
    let policy_text = r#"
action action1() {}
action action2(x int) {
    action action1()
    publish Cmd1 {}
}
action action3() {}

command Cmd1 {
    fields {}
    seal { return None }
    open { return None }
}

effect Effect1 {
    field1 int,
    field2 string,
}
effect Effect2 {}

// TODO: Add enum

fact Fact1[a int, b string]=>{c bool}
fact Fact2[]=>{}

// TODO: add function inputs
finish function func4() {}
finish function func5() {
    func4()
    delete Fact1[a: 42, b: "test"]
}

// TODO: add function inputs
function func1() int { return 42 }
function func2() bool { return true }
function func3() int { return func1() }

struct Struct1 {
    field1 bool,
    field2 int,
}
"#;

    let policy = parse_policy_str(policy_text, Version::V2).unwrap();
    let (hir, ast_nodes) = hir::lower(&policy, &[]);

    // Create snapshot of the parsed HIR and AST nodes
    let snapshot = ParseSnapshot {
        hir: &hir,
        ast_nodes: &ast_nodes,
    };

    insta::assert_json_snapshot!("basic_policy", &snapshot);
}

#[test]
fn test_parse_empty_policy() {
    let policy_text = "";
    let policy = parse_policy_str(policy_text, Version::V2).unwrap();
    let (hir, ast_nodes) = hir::lower(&policy, &[]);

    let snapshot = ParseSnapshot {
        hir: &hir,
        ast_nodes: &ast_nodes,
    };

    insta::assert_json_snapshot!("empty_policy", &snapshot);
}

#[test]
fn test_parse_complex_expressions() {
    let policy_text = r#"
function complex_expr(x int, y int) bool {
    let sum = x + y
    let diff = x - y
    let is_positive = sum > 0
    let is_equal = x == y
    let result = is_positive && !is_equal
    return result || (diff < 0)
}

function optional_handling(x optional int) int {
    return unwrap x
}

function match_expr(x int) string {
    match x {
        1 => { return "one" }
        2 => { return "two" }
        _ => { return "other" }
    }
}
"#;

    let policy = parse_policy_str(policy_text, Version::V2).unwrap();
    let (hir, ast_nodes) = hir::lower(&policy, &[]);

    let snapshot = ParseSnapshot {
        hir: &hir,
        ast_nodes: &ast_nodes,
    };

    insta::assert_json_snapshot!("complex_expressions", &snapshot);
}

#[test]
fn test_parse_nested_structures() {
    let policy_text = r#"
struct Address {
    street string,
    city string,
}

struct Person {
    name string,
    age int,
    address struct Address,
}

command CreatePerson {
    fields {
        person struct Person,
    }
    seal { return None }
    open { return None }
}

action create_person(name string, age int, street string, city string) {
    let addr = Address {
        street: street,
        city: city,
    }
    let person = Person {
        name: name,
        age: age,
        address: addr,
    }
    publish CreatePerson { person: person }
}
"#;

    let policy = parse_policy_str(policy_text, Version::V2).unwrap();
    let (hir, ast_nodes) = hir::lower(&policy, &[]);

    let snapshot = ParseSnapshot {
        hir: &hir,
        ast_nodes: &ast_nodes,
    };

    insta::assert_json_snapshot!("nested_structures", &snapshot);
}

#[test]
fn test_parse_fact_operations() {
    let policy_text = r#"
fact User[user_id string]=>{name string, role string}
fact Permission[user_id string, resource string]=>{allowed bool}

action check_permissions(user_id string) {
    // map is valid in actions
    map Permission[user_id: user_id, resource: ?] as perm {
        check perm.allowed
    }
}

action check_admin(user_id string) {
    let user = query User[user_id: user_id]
    check user.role == "admin"
}

effect PermissionGranted {
    user_id string,
    resource string,
}

finish function grant_permission(user_id string, resource string) {
    create Permission[user_id: user_id, resource: resource]=>{allowed: true}
    // emit is valid in finish functions
    emit PermissionGranted {
        user_id: user_id,
        resource: resource,
    }
}

finish function revoke_permission(user_id string, resource string) {
    delete Permission[user_id: user_id, resource: resource]
}
"#;

    let policy = parse_policy_str(policy_text, Version::V2).unwrap();
    let (hir, ast_nodes) = hir::lower(&policy, &[]);

    let snapshot = ParseSnapshot {
        hir: &hir,
        ast_nodes: &ast_nodes,
    };

    insta::assert_json_snapshot!("fact_operations", &snapshot);
}
