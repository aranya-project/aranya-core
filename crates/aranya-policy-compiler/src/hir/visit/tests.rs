//! Comprehensive tests for the HIR visitor pattern.
//!
//! These tests ensure that all HIR node types are properly visited in the correct order.
//! Each test uses the `RecordingVisitor` to capture the exact sequence of node visits
//! and verifies it against snapshots using `insta`.
//!
//! The tests cover:
//! - All top-level definitions (actions, commands, effects, enums, facts, functions, structs)
//! - All statement types
//! - All expression types
//! - Complex combinations of features
//!
//! ## Test Structure
//!
//! Each test follows the same pattern:
//! 1. Define a policy string with the syntax to test
//! 2. Parse the policy using `parse_policy_str`
//! 3. Generate HIR using `hir::parse`
//! 4. Create a `RecordingVisitor` that captures all visited nodes
//! 5. Walk the HIR and capture all visited nodes
//! 6. Verify the captured nodes against a snapshot using `insta`
//!
//! ## Updating Tests
//!
//! When the HIR structure changes:
//! 1. Run `cargo test -p aranya-policy-compiler hir::visit::tests`
//! 2. Review the changes with `cargo insta review`
//! 3. Accept valid changes with `cargo insta accept`

use aranya_policy_ast::Version;
use aranya_policy_lang::lang::parse_policy_str;
use serde::Serialize;

use super::*;
use crate::hir::{self, hir::*};

/// Enum representing all visitable items in the HIR
#[derive(Clone, Debug, Serialize)]
enum Item {
    ActionDef(ActionDef),
    ActionArg(ActionArg),
    Block(Block),
    CmdDef(CmdDef),
    CmdField(CmdField),
    EffectDef(EffectDef),
    EffectField(EffectField),
    EnumDef(EnumDef),
    Expr(Expr),
    FactDef(FactDef),
    FactKey(FactKey),
    FactVal(FactVal),
    FactLiteral(FactLiteral),
    FinishFuncDef(FinishFuncDef),
    FinishFuncArg(FinishFuncArg),
    FuncDef(FuncDef),
    FuncArg(FuncArg),
    GlobalLetDef(GlobalLetDef),
    Ident(Ident),
    Stmt(Stmt),
    StructDef(StructDef),
    StructField(StructField),
    VType(VType),
}

/// A visitor that records all HIR nodes being visited for snapshot testing
struct RecordingVisitor {
    items: Vec<Item>,
}

impl RecordingVisitor {
    fn new() -> Self {
        Self { items: Vec::new() }
    }
}

impl<'hir> Visitor<'hir> for RecordingVisitor {
    type Result = ();

    fn visit_action_def(&mut self, def: &'hir ActionDef) {
        self.items.push(Item::ActionDef(def.clone()));
    }

    fn visit_action_arg(&mut self, arg: &'hir ActionArg) {
        self.items.push(Item::ActionArg(arg.clone()));
    }

    fn visit_block(&mut self, block: &'hir Block) {
        self.items.push(Item::Block(block.clone()));
    }

    fn visit_cmd_def(&mut self, def: &'hir CmdDef) {
        self.items.push(Item::CmdDef(def.clone()));
    }

    fn visit_cmd_field(&mut self, field: &'hir CmdField) {
        self.items.push(Item::CmdField(field.clone()));
    }

    fn visit_effect_def(&mut self, def: &'hir EffectDef) {
        self.items.push(Item::EffectDef(def.clone()));
    }

    fn visit_effect_field(&mut self, field: &'hir EffectField) {
        self.items.push(Item::EffectField(field.clone()));
    }

    fn visit_enum_def(&mut self, def: &'hir EnumDef) {
        self.items.push(Item::EnumDef(def.clone()));
    }

    fn visit_expr(&mut self, expr: &'hir Expr) {
        self.items.push(Item::Expr(expr.clone()));
    }

    fn visit_fact_def(&mut self, def: &'hir FactDef) {
        self.items.push(Item::FactDef(def.clone()));
    }

    fn visit_fact_key(&mut self, key: &'hir FactKey) {
        self.items.push(Item::FactKey(key.clone()));
    }

    fn visit_fact_value(&mut self, val: &'hir FactVal) {
        self.items.push(Item::FactVal(val.clone()));
    }

    fn visit_finish_func_def(&mut self, def: &'hir FinishFuncDef) {
        self.items.push(Item::FinishFuncDef(def.clone()));
    }

    fn visit_finish_func_arg(&mut self, arg: &'hir FinishFuncArg) {
        self.items.push(Item::FinishFuncArg(arg.clone()));
    }

    fn visit_func_def(&mut self, def: &'hir FuncDef) {
        self.items.push(Item::FuncDef(def.clone()));
    }

    fn visit_func_arg(&mut self, arg: &'hir FuncArg) {
        self.items.push(Item::FuncArg(arg.clone()));
    }

    fn visit_global_def(&mut self, def: &'hir GlobalLetDef) {
        self.items.push(Item::GlobalLetDef(def.clone()));
    }

    fn visit_ident(&mut self, ident: &'hir Ident) {
        self.items.push(Item::Ident(ident.clone()));
    }

    fn visit_stmt(&mut self, stmt: &'hir Stmt) {
        self.items.push(Item::Stmt(stmt.clone()));
    }

    fn visit_struct_def(&mut self, def: &'hir StructDef) {
        self.items.push(Item::StructDef(def.clone()));
    }

    fn visit_struct_field(&mut self, field: &'hir StructField) {
        self.items.push(Item::StructField(field.clone()));
    }

    fn visit_vtype(&mut self, ty: &'hir VType) {
        self.items.push(Item::VType(ty.clone()));
    }

    fn visit_fact_literal(&mut self, fact: &'hir FactLiteral) {
        self.items.push(Item::FactLiteral(fact.clone()));
    }
}

/// Tests visiting an `action`.
#[test]
fn test_action() {
    let text = r#"
action foo(x int, y string) {
    let z = x + 1
}
"#;

    let policy = parse_policy_str(text, Version::V2).unwrap();
    let (hir, _) = hir::parse(&policy, &[]);

    let mut visitor = RecordingVisitor::new();
    hir.walk(&mut visitor);

    insta::assert_json_snapshot!("test_action", &visitor.items);
}

#[test]
fn test_command() {
    let policy_text = r#"
command Foo {
    fields {
        a int,
        b string
    }
    seal { return None }
    open { return None }
    policy {
        finish {
            check true
        }
    }
    recall { return None }
}
    "#;

    let policy = parse_policy_str(policy_text, Version::V2).unwrap();
    let (hir, _) = hir::parse(&policy, &[]);

    let mut visitor = RecordingVisitor::new();
    hir.walk(&mut visitor);

    insta::assert_json_snapshot!("test_command", &visitor.items);
}

#[test]
fn test_effect() {
    let policy_text = r#"
effect MyEffect {
    user_id string,
    timestamp int,
    active bool
}
    "#;

    let policy = parse_policy_str(policy_text, Version::V2).unwrap();
    let (hir, _) = hir::parse(&policy, &[]);

    let mut visitor = RecordingVisitor::new();
    hir.walk(&mut visitor);

    insta::assert_json_snapshot!("test_effect", &visitor.items);
}

#[test]
fn test_enum() {
    let policy_text = r#"
enum Status {
    Active,
    Inactive
}

action use_enum() {
    let s = Status::Active
}
    "#;

    let policy = parse_policy_str(policy_text, Version::V2).unwrap();
    let (hir, _) = hir::parse(&policy, &[]);

    let mut visitor = RecordingVisitor::new();
    hir.walk(&mut visitor);

    insta::assert_json_snapshot!("test_enum", &visitor.items);
}

#[test]
fn test_fact() {
    let policy_text = r#"
fact User[user_id string]=>{name string, age int}

action query_fact() {
    let u = query User[user_id: "123"]
    check exists User[user_id: "456"]
}
    "#;

    let policy = parse_policy_str(policy_text, Version::V2).unwrap();
    let (hir, _) = hir::parse(&policy, &[]);

    // Analysis:
    // - Actions are visited first, then facts
    // - query and exists create InternalFunction expressions
    // - FactLiteral contains identifier and field mappings

    let mut visitor = RecordingVisitor::new();
    hir.walk(&mut visitor);

    insta::assert_json_snapshot!("test_fact", &visitor.items);
}

#[test]
fn test_finish_function() {
    let policy_text = r#"
effect UserDeleted {
    user_id string
}

finish function cleanup(user_id string) {
    delete User[user_id: user_id]
    emit UserDeleted { user_id: user_id }
}
    "#;

    let policy = parse_policy_str(policy_text, Version::V2).unwrap();
    let (hir, _) = hir::parse(&policy, &[]);

    let mut visitor = RecordingVisitor::new();
    hir.walk(&mut visitor);

    insta::assert_json_snapshot!("test_finish_function", &visitor.items);
}

#[test]
fn test_function() {
    let policy_text = r#"
function add(a int, b int) int {
    return a + b
}

function multiply(x int, y int) int {
    let result = x + y
    return result
}
    "#;

    let policy = parse_policy_str(policy_text, Version::V2).unwrap();
    let (hir, _) = hir::parse(&policy, &[]);

    let mut visitor = RecordingVisitor::new();
    hir.walk(&mut visitor);

    insta::assert_json_snapshot!("test_function", &visitor.items);
}

#[test]
fn test_global_let() {
    let policy_text = r#"
let MAX_SIZE = 100
let DEFAULT_NAME = "anonymous"
    "#;

    let policy = parse_policy_str(policy_text, Version::V2).unwrap();
    let (hir, _) = hir::parse(&policy, &[]);

    let mut visitor = RecordingVisitor::new();
    hir.walk(&mut visitor);

    insta::assert_json_snapshot!("test_global_let", &visitor.items);
}

#[test]
fn test_struct() {
    let policy_str = r#"
struct Point {
    x int,
    y int
}

struct Line {
    start struct Point,
    end struct Point,
    color optional string
}
"#;
    let policy = parse_policy_str(policy_str, Version::V2).unwrap();
    let (hir, _) = hir::parse(&policy, &[]);

    let mut visitor = RecordingVisitor::new();
    hir.walk(&mut visitor);

    insta::assert_json_snapshot!("test_struct", &visitor.items);
}

#[test]
fn test_expressions() {
    let policy_str = r#"
function test_expressions() int {
    // Test core expression types
    let a = 42              // Integer literal
    let b = a + 5           // Binary operation
    let c = -a              // Unary operation
    let d = Point { x: 1, y: 2 }  // Struct literal
    let e = d.x             // Field access
    let f = Some(10)        // Optional with value
    let g = unwrap f        // Unwrap operation

    return g
}

struct Point {
    x int,
    y int
}
"#;
    let policy = parse_policy_str(policy_str, Version::V2).unwrap();
    let (hir, _) = hir::parse(&policy, &[]);

    let mut visitor = RecordingVisitor::new();
    hir.walk(&mut visitor);

    insta::assert_json_snapshot!("test_expressions", &visitor.items);
}

#[test]
fn test_statements() {
    let policy_str = r#"
action test_statements() {
    // Test check statement
    check true

    // Test if statement
    if true {
        let x = 1
    } else {
        let x = 2
    }

    // Test match statement
    let status = 1
    match status {
        1 => {
            let result = "one"
        }
        2 => {
            let result = "two"
        }
        _ => {
            let result = "other"
        }
    }

    // Test create statement
    create User[user_id: "test"]=>{ name: "Test User" }

    // Test update statement
    update User[user_id: "test"]=>{ name: ? } to { name: "Updated User" }

    // Test delete statement
    delete User[user_id: "test"]

    // Test emit statement
    emit UserUpdated { user_id: "test" }

    // Test publish statement
    publish TestCommand { value: 42 }
}

fact User[user_id string]=>{ name string }

effect UserUpdated {
    user_id string
}

command TestCommand {
    fields {
        value int
    }
    seal {}
    open {}
    policy {}
    recall { return None }
}
"#;
    let policy = parse_policy_str(policy_str, Version::V2).unwrap();
    let (hir, _) = hir::parse(&policy, &[]);

    let mut visitor = RecordingVisitor::new();
    hir.walk(&mut visitor);

    insta::assert_json_snapshot!("test_statements", &visitor.items);
}

#[test]
fn test_struct_refs() {
    let policy_str = r#"
struct Point {
    x int,
    y int,
}

struct Size {
    width int,
    height int,
}

// Struct that references other structs
struct Rectangle {
    +Point,
    +Size,
}

// Struct with both fields and references
struct NamedRectangle {
    name string,
    +Point,
    +Size,
    color string,
}
"#;
    let policy = parse_policy_str(policy_str, Version::V2).unwrap();
    let (hir, _) = hir::parse(&policy, &[]);

    let mut visitor = RecordingVisitor::new();
    hir.walk(&mut visitor);

    insta::assert_json_snapshot!("test_struct_refs", &visitor.items);
}
