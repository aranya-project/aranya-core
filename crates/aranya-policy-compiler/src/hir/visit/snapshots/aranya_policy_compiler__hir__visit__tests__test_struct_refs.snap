---
source: crates/aranya-policy-compiler/src/hir/visit/tests.rs
expression: "&visitor.items"
---
[
  {
    "StructDef": {
      "id": {
        "idx": 1,
        "version": 1
      },
      "span": {
        "start": 1,
        "end": 1
      },
      "ident": {
        "idx": 1,
        "version": 1
      },
      "items": [
        {
          "idx": 1,
          "version": 1
        },
        {
          "idx": 2,
          "version": 1
        }
      ]
    }
  },
  {
    "StructField": {
      "id": {
        "idx": 1,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "kind": {
        "Field": {
          "ident": {
            "idx": 2,
            "version": 1
          },
          "ty": {
            "idx": 1,
            "version": 1
          }
        }
      }
    }
  },
  {
    "Ident": {
      "id": {
        "idx": 2,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "ident": "x"
    }
  },
  {
    "VType": {
      "id": {
        "idx": 1,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "kind": "Int"
    }
  },
  {
    "StructField": {
      "id": {
        "idx": 2,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "kind": {
        "Field": {
          "ident": {
            "idx": 3,
            "version": 1
          },
          "ty": {
            "idx": 2,
            "version": 1
          }
        }
      }
    }
  },
  {
    "Ident": {
      "id": {
        "idx": 3,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "ident": "y"
    }
  },
  {
    "VType": {
      "id": {
        "idx": 2,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "kind": "Int"
    }
  },
  {
    "StructDef": {
      "id": {
        "idx": 2,
        "version": 1
      },
      "span": {
        "start": 41,
        "end": 41
      },
      "ident": {
        "idx": 4,
        "version": 1
      },
      "items": [
        {
          "idx": 3,
          "version": 1
        },
        {
          "idx": 4,
          "version": 1
        }
      ]
    }
  },
  {
    "StructField": {
      "id": {
        "idx": 3,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "kind": {
        "Field": {
          "ident": {
            "idx": 5,
            "version": 1
          },
          "ty": {
            "idx": 3,
            "version": 1
          }
        }
      }
    }
  },
  {
    "Ident": {
      "id": {
        "idx": 5,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "ident": "width"
    }
  },
  {
    "VType": {
      "id": {
        "idx": 3,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "kind": "Int"
    }
  },
  {
    "StructField": {
      "id": {
        "idx": 4,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "kind": {
        "Field": {
          "ident": {
            "idx": 6,
            "version": 1
          },
          "ty": {
            "idx": 4,
            "version": 1
          }
        }
      }
    }
  },
  {
    "Ident": {
      "id": {
        "idx": 6,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "ident": "height"
    }
  },
  {
    "VType": {
      "id": {
        "idx": 4,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "kind": "Int"
    }
  },
  {
    "StructDef": {
      "id": {
        "idx": 3,
        "version": 1
      },
      "span": {
        "start": 129,
        "end": 129
      },
      "ident": {
        "idx": 7,
        "version": 1
      },
      "items": [
        {
          "idx": 5,
          "version": 1
        },
        {
          "idx": 6,
          "version": 1
        }
      ]
    }
  },
  {
    "StructField": {
      "id": {
        "idx": 5,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "kind": {
        "StructRef": {
          "idx": 8,
          "version": 1
        }
      }
    }
  },
  {
    "Ident": {
      "id": {
        "idx": 8,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "ident": "Point"
    }
  },
  {
    "StructField": {
      "id": {
        "idx": 6,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "kind": {
        "StructRef": {
          "idx": 9,
          "version": 1
        }
      }
    }
  },
  {
    "Ident": {
      "id": {
        "idx": 9,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "ident": "Size"
    }
  },
  {
    "StructDef": {
      "id": {
        "idx": 4,
        "version": 1
      },
      "span": {
        "start": 216,
        "end": 216
      },
      "ident": {
        "idx": 10,
        "version": 1
      },
      "items": [
        {
          "idx": 7,
          "version": 1
        },
        {
          "idx": 8,
          "version": 1
        },
        {
          "idx": 9,
          "version": 1
        },
        {
          "idx": 10,
          "version": 1
        }
      ]
    }
  },
  {
    "StructField": {
      "id": {
        "idx": 7,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "kind": {
        "Field": {
          "ident": {
            "idx": 11,
            "version": 1
          },
          "ty": {
            "idx": 5,
            "version": 1
          }
        }
      }
    }
  },
  {
    "Ident": {
      "id": {
        "idx": 11,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "ident": "name"
    }
  },
  {
    "VType": {
      "id": {
        "idx": 5,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "kind": "String"
    }
  },
  {
    "StructField": {
      "id": {
        "idx": 8,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "kind": {
        "StructRef": {
          "idx": 12,
          "version": 1
        }
      }
    }
  },
  {
    "Ident": {
      "id": {
        "idx": 12,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "ident": "Point"
    }
  },
  {
    "StructField": {
      "id": {
        "idx": 9,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "kind": {
        "StructRef": {
          "idx": 13,
          "version": 1
        }
      }
    }
  },
  {
    "Ident": {
      "id": {
        "idx": 13,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "ident": "Size"
    }
  },
  {
    "StructField": {
      "id": {
        "idx": 10,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "kind": {
        "Field": {
          "ident": {
            "idx": 14,
            "version": 1
          },
          "ty": {
            "idx": 6,
            "version": 1
          }
        }
      }
    }
  },
  {
    "Ident": {
      "id": {
        "idx": 14,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "ident": "color"
    }
  },
  {
    "VType": {
      "id": {
        "idx": 6,
        "version": 1
      },
      "span": {
        "start": 0,
        "end": 0
      },
      "kind": "String"
    }
  }
]
